test/test_session.cpp

```diff
@@ -394,3 +394,73 @@ TORRENT_TEST(save_state_peer_id)
 	TEST_CHECK(pid3[5] == 'r');
 }
 
+TORRENT_TEST(init_dht)
+{
+	auto count_dht_inits = [](session& ses)
+	{
+		int count = 0;
+		int num = 70; // this number is adjusted per version, an estimate
+		time_point const end_time = clock_type::now() + seconds(15);
+		while (true)
+		{
+			time_point const now = clock_type::now();
+			if (now > end_time) return count;
+
+			ses.wait_for_alert(end_time - now);
+			std::vector<alert*> alerts;
+			ses.pop_alerts(&alerts);
+			for (auto a : alerts)
+			{
+				std::printf("%d: [%s] %s\n", num, a->what(), a->message().c_str());
+				if (a->type() == log_alert::alert_type)
+				{
+					std::string const msg = a->message();
+					if (msg.find("about to start DHT") != std::string::npos)
```

Comment:
@aldenml I don't think this test is quite right. it looks for the "about to start DTH" log, but looking at the code where that's emitted, that message is printed even when starting the DHT is suppressed/delayed because there are outstanding lookups. see [here](https://github.com/arvidn/libtorrent/blob/master/src/session_impl.cpp#L5716).

Reply:
That’s is true if update_dht is called like it was before, or like it is in RC_1_1, but the new init_dht here take care of the special cases.

Reply:
I bring this up as I've been merging RC_1_1 into master, with the changes to how ``m_settings`` are initialized. The test for this patch failed, but if I move the log to when the DHT is *actually* started, they all pass. As far as I can tell, the intention of the test was to make sure the DHT is only constructed and started once on startup, regardless of those settings.

As far as I can tell, the merged master (which I still only have in my working copy) doesn't need the ``init_dht()`` function. what happens is that the router nodes are set first, and they kick off a host name lookup. Even if this is a cache hit, it won't return "in-line", it will post the result to the io_service and effectively defer starting the DHT. Next, ``update_dht()`` is called, but it won't actually construct or start the DHT, since there's at least one outstanding lookup.

Once the lookup comes back, the DHT will be started. Only once.

The reason this works in my merged version is because ``update_dht_bootstrap_nodes()`` is called before ``update_dht()``, which is an implementation detail of ``run_all_updates()``, but at least these tests would catch it if it was broken.

Trying to understand all the cases in this patch, I can't think of a way if a case that would break the single-dht-init.

I have not gone back to look at current master to see what would happen there if I moved the log call.